Documentation and release prep (#41)
puppetlabs/puppetlabs-puppet_operations_appliance: README.md

### Setup Requirements 

Module Dependencies

>  - derdanne/nfs (>= 2.1.5)

Reviewer: The formatting is inconsistent with these new ones.

```suggestion
- derdanne/nfs (>= 2.1.5)
```

Reviewer: To solve the formatting issue, you may need to add the `  -` to these.

Reviewer: Nevermind. It would just be one or the other. The unordered list works well.

- herculesteam/augeasproviders_core (>= 2.1.5 < 4.0.0)
- herculesteam/augeasproviders_shellvar (>= 1.2.0 < 5.0.0)
- puppetlabs/apt (>= 2.0.0 < 8.0.0)
- puppet-grafana (>= 3.0.0 < 7.0.0)
- puppet-telegraf (>= 2.0.0 < 4.0.0)
- puppetlabs-apt (>= 4.3.0 < 8.0.0)
- puppetlabs-inifile (>= 2.0.0 < 5.0.0)
- puppetlabs-puppetserver_gem (>= 1.1.1 < 3.0.0)

### Beginning with rsan

RSAN has two main classes for use in the installation:

 - rsan::exporter - to be applied to all Puppet infrastructure agents - Console node group "PE Infrastructure Agent"
 - rsan::importer - to be applied to a single node which will be come the Remote Support Access Node(RSAN)

Following the application of these clases to the infrastructure, Puppet Will need to be run on the corresponding agents in the following order:

Infrastructure Agent(s)->RSAN Agent->Infrastrcture Agent(s)->RSAN Agent

## Usage
The following outlines the main features of RSAN and how to consume them
### Live Telemetry Display
<<<<<<< HEAD

The Rsan node will host an instance of the [Puppet Metrics Dashboard](https://forge.puppet.com/modules/puppetlabs/puppet_metrics_dashboard)
 
The Dashboard can be accessed on

> <RSAN-ip\>:3000\

Reviewer: ```suggestion
http://<RSAN-ip\>:3000\
```

> http://<RSAN-ip\>:3000\

Reviewer: ```suggestion
**URL:** http://<RSAN-ip\>:3000\
```

> User: admin\

Reviewer: ```suggestion
**User:** admin\
```

> Password: admin

Reviewer: ```suggestion
**Password:** admin
```

For advanced configuration and documentation please see [Puppet Metrics Dashboard](https://forge.puppet.com/modules/puppetlabs/puppet_metrics_dashboard)

### Infrastructure node file and log access	

> The RSAN node will, by default, mount /var/log/ /opt/puppetlabs and /etc/puppetlabs from each of the Puppet Enterprise Infrastructure nodes on the RSAN platform in the following location, as read only file systems.

Reviewer: ```suggestion
The RSAN node will, by default, mount `/var/log/`, `/opt/puppetlabs` and `/etc/puppetlabs` from each of the Puppet Enterprise Infrastructure nodes on the RSAN platform in the following location, as read-only file systems.
```

> /var/pesupport/<FQDN of Infrastructure node\>/var/log\

Reviewer: ```suggestion
`/var/pesupport/<FQDN of Infrastructure node\>/var/log\`
```

> /var/pesupport/<FQDN of Infrastructure node\>/opt/puppetlabs\

Reviewer: ```suggestion
`/var/pesupport/<FQDN of Infrastructure node\>/opt/puppetlabs\`
```

> /var/pesupport/<FQDN of Infrastructure node\>/etc/puppetlabs

Reviewer: ```suggestion
`/var/pesupport/<FQDN of Infrastructure node\>/etc/puppetlabs`
```

#### Optional Configuration

The RSAN Class assumes the RSAN server will mount the shared partitions using the IP address Source designated by the "ipaddress" fact. In any deployment should this assertion not be true, it is nessary to set the following parameter to the source IP address of the RSAN Host:

> **rsan::exporter::rsan_importer_ips**

Reviewer: This should show the syntax of setting this parameter or link to the REFERENCE.md for information.

In hiera"

```
rsan::exporter::rsan_importer_ips:
  - 1.2.3.4
 ```

In the class declaration or in the console the value would be an array

```
["1.2.3.4"]
```

For imformation on executing PE tasks see the [Puppet Enterprise Documentation](https://puppet.com/docs/pe/2019.8/tasks_in_pe.html)\
RSAN::supportuser\
When successful the task will return a password, this should be delivered to Puppet Enterprise Support personnel.

The Task creates the following user and role:

**User:** pesupport 

**Role:** PE Suport Role 

The role is intentonally left without permissions, and should be given only the permissions the installing organisation are authorised to grant to Puppet Enterprise Support personnel. For more information on RBAC permissions please see the [Puppet Enterprise Documentation](https://puppet.com/docs/pe/2019.8/rbac_permissions_intro.html)

### Puppet Enterprise Database Access	

The RSAN Platform has a Postgresql client installed, and is granted certificate based access to all Puppet Enterprise Databases on any pe_postgresl node within the current deployment. The access is limited to the [SELECT](https://www.postgresql.org/docs/11/sql-grant.html) privilege and is therefore READONLY in nature.

To use this function execute the following command from the CLI of the RSAN host

```

> psql "sslmode=verify-ca sslrootcert=/etc/puppetlabs/puppet/ssl/certs/ca.pem  sslcert=/etc/puppetlabs/puppet/ssl/certs/<fqdn_rsan_host>.pem sslkey=/etc/puppetlabs/puppet/ssl/private_keys/<fqdn_rsan_host>.pem hostaddr=<pe_postgres_host> port=5432 user=rsan dbname=<pe_db_name>"

Reviewer: For copy/paste ability

```suggestion
psql "host=$(puppet config print server) port=5432 user=rsan sslmode=verify-full sslcert=$(puppet config print hostcert) sslkey=$(puppet config print hostprivkey) sslrootcert=$(puppet config print localcacert) dbname=<pe_db_name>"
```

```

Where valid options for <pe_db_name> are:

- pe-rbac 
- pe-puppetdb 
- pe-orchestrator 
- pe-inventory 
- pe-classifier 
- pe-activity

## Uninstallation 

To Uninsuall RSAN from your Puppet Enterprise Infrastructure.

 - Remove the following Classification:
rsan::exporter\
rsan::importer

 - Add the following classification to the "PE Infrastructure Agent" node group
 rsan::remove_exporter

  - Remove the following classification to the "PE Infrastructure Agent" node group
 rsan::remove_exporter

  - Run Puppet on all nodes in "PE Infrastructure Agent" node group

   - Decommission the RSAN platform 

## Limitations
 - The RSAN importer class should only be applied one agent node

>  - All features are currently 

Reviewer: Currently what?

Author: doh!

## Known Issues

- When accessing the database from the RSAN node for the first time, an error message will be presented: [#40](https://github.com/MartyEwings/RSAN/issues/40)

```psql: private key file "/etc/puppetlabs/puppet/ssl/private_keys/rsan-target.platform9.puppet.net.pem" has group or world access; permissions should be u=rw (0600) or less```

>  To workaround change the private key file to 0600 or less, Puppet will manage this file back to incorrect permissions, for long term access making a copy of the key with the correct permissions is the best course of action. 

Reviewer: Potential workaround: https://puppet.com/docs/puppet/7.1/config_file_main.html#special-types-of-values-for-settings

> Settings that take a single file or directory (like ssldir) can accept an optional hash of permissions. When starting up, Puppet enforces those permissions on the file or directory.

Perhaps we can set that for the `privatekeydir` or `hostprivkey` setting.

Author: im going to see if i can resolve this in an issue

#40

 - PuppetDB Metric Collection fails due to CVE-2020-7943  [27](https://github.com/MartyEwings/RSAN/issues/27)

> Please refer to the documenation of Puppet Metrics collector for recommended work arounds

Reviewer: This would be the puppet metrics dashboard module for this deployment.

Author: doh

Reviewer: ```suggestion
Please refer to the documentation of [Puppet Metrics Dashboard](https://forge.puppet.com/modules/puppetlabs/puppet_metrics_dashboard) for recommended workarounds
```

`/var/pesupport/<FQDN of Infrastructure node\>/var/log\`
`/var/pesupport/<FQDN of Infrastructure node\>/opt/puppetlabs\`
`/var/pesupport/<FQDN of Infrastructure node\>/etc/puppetlabs`

#### Optional Configuration

The RSAN Class assumes the RSAN server will mount the shared partitions using the IP address Source designated by the "ipaddress" fact. In any deployment should this assertion not be true, it is nessary to set the following parameter to the source IP address of the RSAN Host:

**rsan::exporter::rsan_importer_ips**

### PE Client tools

The RSAN node will deploy Puppet Client tools for use by Puppet Enterprise on the RSAN platform, For More information please see the Puppet Enterprise Documentation:

[PE Client tools](https://puppet.com/docs/pe/2019.8/installing_pe_client_tools.html)

A supplementary task is available to generate an RBAC user and role, so that the credentials may be used provided to Puppet Enterprise Support personnel.

#### Creating Support User

Run the following task against the Primary Puppet Enterprise Server\

> For imformation on executing PE tasks see the [Puppet Enterprise Documentation](https://puppet.com/docs/pe/2019.8/tasks_in_pe.html)\

Reviewer: ```suggestion
For information on executing PE tasks see the [Puppet Enterprise Documentation](https://puppet.com/docs/pe/2019.8/tasks_in_pe.html)\
```

 - rsan::importer - to be applied to a single node which will be come the Remote Support Access Node(RSAN)

Following the application of these clases to the infrastructure, Puppet Will need to be run on the corresponding agents in the following order:

Infrastructure Agent(s)->RSAN Agent->Infrastrcture Agent(s)->RSAN Agent

## Usage
The following outlines the main features of RSAN and how to consume them
### Live Telemetry Display

The Rsan node will host an instance of the [Puppet Metrics Dashboard](https://forge.puppet.com/modules/puppetlabs/puppet_metrics_dashboard)
 
The Dashboard can be accessed on

**URL:** http://<RSAN-ip\>:3000\
**User:** admin\
**Password:** admin

For advanced configuration and documentation please see [Puppet Metrics Dashboard](https://forge.puppet.com/modules/puppetlabs/puppet_metrics_dashboard)

### Infrastructure node file and log access	

The RSAN node will, by default, mount `/var/log/`, `/opt/puppetlabs` and `/etc/puppetlabs` from each of the Puppet Enterprise Infrastructure nodes on the RSAN platform in the following location, as read-only file systems.

> `/var/pesupport/<FQDN of Infrastructure node\>/var/log\`

Reviewer: My bad on these. They need to be an unordered list and remove the `\>` now that they are in the inline code.

```
* `/var/pesupport/<FQDN of Infrastructure node>/var/log`
```
